### PR TITLE
[WIP] - Directory Structure

### PR DESCRIPTION
**Pull request by seanpmorgan**
**This PR is not ready to be merged.**

Creating this PR early so we can review the initial directory structure & build setup. Branch still needs modifications to match establish API patterns and ensure tests are passing.

### CODE REVIEW

**Comment by armando-fandango** [comment]
- tensorflow_addons could be just named addons
- opt: would optimizers or optimization make it more clear ?
- metrics : for functions to use as loss function or evaluation metrics

**Comment by seanpmorgan** [comment]
@armando-fandango @karmel Updated the directory structure and built working tests. addons.text compiles fine and whl packaging works as well.

**Comment by armando-fandango** [comment]
Thats awesome, great work and congrats !!!!

Armando Fandango

*Links to my* books

Deep Learning, TensorFlow and Keras: *http://amzn.to/2tFKdOW
<http://amzn.to/2tFKdOW>*
Machine Learning: https://amzn.to/2Qfnm7S
Data Science: *http://amzn.to/2FQayib <http://amzn.to/2FQayib>*

File: README.md [hunk 1]
```diff
@@ -16,9 +16,18 @@ The tensorflow/addons repository, will contain additional functionality fitting
...
----
+# Developing
 
-<br><br>
-<div align="center">
-    <img src ="https://cdn.pixabay.com/photo/2017/06/16/07/26/under-construction-2408062_640.png" />
-</div>
+## Docker
+**Note:** This docker container is just temporary until we can pull a
+tensorflow/tensorflow:custom-op container that reflects nightly changes.
```
Comment (karmel) [inline README.md]: @yifeif - This is less than ideal; do we have an ETA on the nightly custom op build?

File: setup.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
+# Copyright 2018 The TensorFlow Authors. All Rights Reserved.
```
Comment (karmel) [inline setup.py]: nit: 2019, just barely

File: tensorflow_addons/crf/README.md [hunk 3]
```diff
@@ -0,0 +1,4 @@
+# Addons - Conditional Random Fields (CRF)
+
+
+## Standard API
```
Comment (karmel) [inline tensorflow_addons/crf/README.md]: Github isn't making it easy to comment on empty files, so commenting here: if we don't need the python subdirs here (ie, crf/python/__init__.py), might be easier/clearer to remove throughout? Unless you feel it is helpful to have python/ versus cc/ for ops.

File: tensorflow_addons/layers/README.md [hunk 4]
```diff
@@ -0,0 +1,6 @@
+# Addons - Layers
+
+
+## Standard API
+In order to conform with the current API standard, all layers
+must inherit from either `keras.layers.Layer` or it's subclasses.
```
Comment (karmel) [inline tensorflow_addons/layers/README.md]: nit: its subclasses

File: tf/BUILD.tpl [hunk 5]
```diff
@@ -0,0 +1,18 @@
+package(default_visibility = ["//visibility:public"])
```
Comment (karmel) [inline tf/BUILD.tpl]: Is the tf/ dir intended strictly for TensorFlow deps? Perhaps a more explicit name -- or a more general one if we anticipate needing other deps here eventually -- would help people navigate?

File: tensorflow_addons/crf/README.md (hunk 3 above)
Comment (seanpmorgan) [reply tensorflow_addons/crf/README.md]: IMO I like the separation between python vs. cc, but am open to removing it. As far as the entirely empty dirs (like crf), the init is just there so the directory structure can be checked into git.

File: tf/BUILD.tpl (hunk 5 above)
Comment (seanpmorgan) [reply tf/BUILD.tpl]: The directory name was just copied from [custom-op](https://github.com/tensorflow/custom-op). I agree though it's not really describing what it does. Perhaps `tf_dependency`?

**Comment by seanpmorgan** [comment]
@karmel The only changes other than directory structure wasthis commit https://github.com/tensorflow/addons/pull/2/commits/f7f2381cbc2586b99988fb3a7caaeb9cfdf21393 to make it so the tests would run in TF2 preview. 

There was one test I couldn't get to work, and I intended to create a bug issue after this merge to discuss. The test was for throwing an error when skip gram tensor was not rank 1. Currently I I just decorated that test with `@test_util.run_deprecated_v1` until it could be addressed.

File: tensorflow_addons/layers/python/wrappers.py [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+    conditioning of the optimization problem.
+    Weight Normalization: A Simple Reparameterization to Accelerate
+    Training of Deep Neural Networks: https://arxiv.org/abs/1602.07868
+    Tim Salimans, Diederik P. Kingma (2016)
+    WeightNorm wrapper works for keras and tf layers.
+    ```python
+      net = WeightNorm(tf.keras.layers.Conv2D(2, 2, activation='relu'),
+             input_shape=(32, 32, 3), data_init=True)(x)
+      net = WeightNorm(tf.keras.layers.Conv2D(16, 5, activation='relu'),
+                       data_init=True)
```
Comment (erfannoury) [inline tensorflow_addons/layers/python/wrappers.py]:
```suggestion
                       data_init=True)(net)
```

File: tensorflow_addons/layers/python/wrappers.py [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+        self.initialized = True
+        if data_init:
+            self.initialized = False
+
+        super(WeightNorm, self).__init__(layer, **kwargs)
+        self._track_checkpointable(layer, name='layer')
+
+    def _compute_weights(self):
+        """Generate weights by combining the direction of weight vector
+         with it's norm """
```
Comment (erfannoury) [inline tensorflow_addons/layers/python/wrappers.py]:
```suggestion
         with its norm """
```

File: README.md (hunk 1 above)
Comment (yifeif) [reply README.md]: I just pushed the first version of the nightly at tensorflow/tensorflow:nightly-custom-op. Hoping to get the nightly push setup by end of week.

Comment (seanpmorgan) [reply README.md]: Thanks!

File: tf/BUILD.tpl (hunk 5 above)
Comment (karmel) [reply tf/BUILD.tpl]: Ah, if it's copied over, probably best to leave as-is for now for consistency. @yifeif - maybe a more explicit name for that would help?

File: tensorflow_addons/crf/README.md (hunk 3 above)
Comment (karmel) [reply tensorflow_addons/crf/README.md]: SGTM

File: tf/BUILD.tpl (hunk 5 above)
Comment (karmel) [reply tf/BUILD.tpl]: Although now I see you have updated it already; either way is fine by me.
